src/simulator/halo_simulator.py

```diff
@@ -223,17 +237,59 @@ def simulated_venn_diagram_reach_by_spend(
         Args:
             spends:  The hypothetical spend vector, equal in length to
               the number of publishers.  spends[i] is the amount that is
-              spent with publisher i.
+              spent with publisher i. Note that publishers with 0 spends will
+              not be included in the Venn diagram reach.
             budget:  The amount of privacy budget that can be consumed while
               satisfying the request.
             max_frequency:  The maximum frequency for which to report reach.
         Returns:
-            A list of ReachPoint. Each reach point represents the mapping from 
+            A list of ReachPoint. Each reach point represents the mapping from
             the spends of a subset of publishers to the differentially private
             estimate of the number of people reached in this subset.
         """
         raise NotImplementedError()
 
+    def _form_venn_diagram_regions(self, spends: List[float]):
+        """Form Venn diagram regions that contain frequency"""
+        # Get user counts by spend for each active publisher
+        user_counts_by_pub_id = {}
+        for pub_id, spend in enumerate(spends):
+            if not spend:
+                continue
+            user_counts_by_pub_id[pub_id] = self._publishers[
+                pub_id
+            ]._publisher_data.user_counts_by_spend(spend)
+
+        if len(user_counts_by_pub_id) > MAX_ACTIVE_PUBLISHERS:
+            raise ValueError("Too many active publishers for Venn diagram algorithm.")
+
+        # Locate user's region and sum the impressions. Also, map the region
+        # representation to publisher ids.
+        user_info = collections.defaultdict(UserInfo)
+        pub_set_by_region = collections.defaultdict(set)
+
+        for pub_id, user_counts in user_counts_by_pub_id.items():
+            for user_id, impressions in user_counts.items():
+                lr = user_info[user_id].located_region
+                new_lr = lr | (1 << pub_id)
+
+                if not new_lr in pub_set_by_region:
+                    pub_set_by_region[new_lr] = copy.deepcopy(pub_set_by_region[lr])
+                    pub_set_by_region[new_lr].add(pub_id)
+
+                user_info[user_id].located_region = new_lr
+                user_info[user_id].impressions += impressions
+
+        # Fill in 2^p - 1 regions of the Venn diagram. Note that the region with
+        # representation 0 is unused
+        num_publishers = len(spends)
+        regions = [{} for _ in range(2 ** num_publishers)]
+        for user_id, user_info in user_info.items():
+            lr = user_info.located_region
+            regions[lr][user_id] = user_info.impressions
+
+        return pub_set_by_region, regions
```

Comment:
@matthewclegg @jiayu-google 
I know we have said to compute the `reach` (number of unique users) in the regions, but I wonder if obtaining more information than just `reach` will be helpful. 

For example, the code above obtains regions such that every region contains `frequency` info represented by a dictionary of `user_id` to number of impressions. This might be useful later on when we compute **impressions of publishers** and **k+ reaches** for initializing a reach point.

If that's not what we want, it's easy to switch back to computing `reach` by counting the number of users in a region.

Please let me know what you think. Thanks!

Reply:
Based on the discussion with @matthewclegg , we will only compute `reach` in the regions. As a result, the k+ reaches will be 1+ reaches. The impressions of publishers can be directly computed from the spend vector.